api/provider.go

```diff
@@ -0,0 +1,190 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/bitbucket"
+	"golang.org/x/oauth2/github"
+)
+
+// Provider is an interface for interacting with external account providers
+type Provider interface {
+	GetUserEmail(context.Context, *oauth2.Token) (string, error)
+	GetOAuthToken(context.Context, string) (*oauth2.Token, error)
+}
+
+var knownProviders = map[string]func(key, secret string) Provider{
+	"github":    NewGithubProvider,
+	"bitbucket": NewBitbucketProvider,
+}
+
+// Provider returns a Provider inerface for the given name
+func (a *API) Provider(name string) (Provider, error) {
+	fn, ok := knownProviders[name]
+	if !ok {
+		return nil, errors.New("Unknown provider")
+	}
+
+	name = strings.ToLower(name)
+
+	var key, secret string
+	switch name {
+	case "github":
+		key, secret = a.config.External.Github.Key, a.config.External.Github.Secret
+	case "bitbucket":
+		key, secret = a.config.External.Bitbucket.Key, a.config.External.Bitbucket.Secret
+	}
+
+	if key == "" || secret == "" {
```

Comment:
instead of this couldn't you just add a default case? And couldn't you combine this with the check above for knownProviders? 
```
switch name {
case "github": 
  return NewGithubProvider(........), nil
case "bitbucket":
  return NewBitbucketProvider(......), nil
}

return nil, errors.Errorf("Unknown provider %s", name)
```

The indirection in the map above doesn't seem to be adding anything